c_src/erocksdb.cc

```
@@ -1176,31 +1459,96 @@ erocksdb_open(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
 
-    // create a resource reference to send erlang
+
+} // erocksdb_open
+
+ERL_NIF_TERM
+erocksdb_open_with_cf(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
+{
+    char db_name[4096];
+    erocksdb::DbObject * db_ptr;
+    rocksdb::DB *db(0);
+
+    if(!enif_get_string(env, argv[0], db_name, sizeof(db_name), ERL_NIF_LATIN1) ||
+       !enif_is_list(env, argv[1]) || !enif_is_list(env, argv[2]))
+    {
+        return enif_make_badarg(env);
+    }
+
+    // parse main db options
+    rocksdb::Options *opts = new rocksdb::Options;
+    fold(env, argv[1], parse_db_option, *opts);
+
+    std::vector<rocksdb::ColumnFamilyDescriptor> column_families;
+
+    ERL_NIF_TERM head, tail = argv[2];
+    while(enif_get_list_cell(env, tail, &head, &tail))
+    {
+        ERL_NIF_TERM result = parse_cf_descriptor(env, head, column_families);
+        if (result != erocksdb::ATOM_OK)
+        {
+            return result;
+        }
+    }
+
+    std::vector<rocksdb::ColumnFamilyHandle*> handles;
+    rocksdb::Status status = rocksdb::DB::Open(*opts, db_name, column_families, &handles, &db);
+
+    if(!status.ok())
+        return error_tuple(env, erocksdb::ATOM_ERROR_DB_OPEN, status);
+
+    // create db respirce
+    db_ptr = erocksdb::DbObject::CreateDbObject(db, opts);
     ERL_NIF_TERM result = enif_make_resource(env, db_ptr);
+
+    unsigned int num_cols;
+    enif_get_list_length(env, argv[2], &num_cols);
+
+    ERL_NIF_TERM cf_list = enif_make_list(env, 0);
+    try {
+        for (int i = 0; i < num_cols; ++i)
+        {
+            erocksdb::ColumnFamilyObject * handle_ptr;
+            handle_ptr = erocksdb::ColumnFamilyObject::CreateColumnFamilyObject(db_ptr, handles[i]);
+            ERL_NIF_TERM cf = enif_make_resource(env, handle_ptr);
+            enif_release_resource(handle_ptr);
+            handle_ptr = NULL;
+            cf_list = enif_make_list_cell(env, cf, cf_list);
+        }
+    } catch (const std::exception& e) {
+        // pass through
```

why not return the error tuple rather than `pass through`?